widget/textarea.go

```diff
@@ -32,6 +32,8 @@ type TextArea struct {
 	vSlider         *Slider
 	hSlider         *Slider
 	text            *Text
+	MaxWidth        int
```

Comment:
Is there a reason you cannot use the RowLayout or GridLayout MaxWidth / MaxHeight to control this?

I can be convinced that we should allow widgets to set their maxwidth and maxheight if there is a good use case, but if we do it should be in the widget object not here. (See MinWidth and MinHeight).

Reply:
Hey, I did try to use the RowLayout MaxWidth and MaxHeight, but that did not took effect, did try different ways but the textarea always went bigger, will try a bit more, may be I can share that code here. But will have to get back later.

Reply:
```
package ui

import (
	"fmt"
	"moiten/archetype"
	"moiten/assets"
	"moiten/component"
	"time"

	"github.com/ebitenui/ebitenui"
	"github.com/ebitenui/ebitenui/image"
	"github.com/ebitenui/ebitenui/widget"
	"github.com/hajimehoshi/ebiten/v2"
	"github.com/yohamta/donburi"
	"golang.org/x/image/colornames"
)

const (
	sendPlaceHolder = "Send"
)

type listResources struct {
	image        *widget.ScrollContainerImage
	track        *widget.SliderTrackImage
	trackPadding widget.Insets
	handle       *widget.ButtonImage
	handleSize   int
	entryPadding widget.Insets
}

type ActionsBar struct {
	container    *widget.Container
	ui           *ebitenui.UI
	closeButton  *widget.Button
	shareButton  *widget.Button
	textArea     *widget.TextArea
	Close        bool
	Share        bool
	Game         *component.GameData
	World        *donburi.World
	MessageValue *string
	MessageDone  bool
	inputText    *widget.TextInput
	sendButton   *widget.Button
}

func NewActionsBar() *ActionsBar {

	container := widget.NewContainer(
		widget.ContainerOpts.Layout(widget.NewAnchorLayout()),
	)
	actionsBar := &ActionsBar{}

	compositeContainer := widget.NewContainer(
		widget.ContainerOpts.Layout(widget.NewRowLayout(
			widget.RowLayoutOpts.Spacing(0),
			widget.RowLayoutOpts.Direction(widget.DirectionVertical),
			widget.RowLayoutOpts.Padding(widget.NewInsetsSimple(0)))),
		widget.ContainerOpts.WidgetOpts(widget.WidgetOpts.LayoutData(widget.AnchorLayoutData{
			HorizontalPosition: widget.AnchorLayoutPositionStart,
			VerticalPosition:   widget.AnchorLayoutPositionStart,
			StretchHorizontal:  false,
			StretchVertical:    false,
		})),
	)
	container.AddChild(compositeContainer)

	textAreaContainer := widget.NewContainer(
		widget.ContainerOpts.WidgetOpts(widget.WidgetOpts.LayoutData(widget.RowLayoutData{
			Stretch:   false,
			MaxWidth:  200,
			MaxHeight: 600,
		})),
		widget.ContainerOpts.Layout(widget.NewRowLayout(
			widget.RowLayoutOpts.Spacing(0),
			widget.RowLayoutOpts.Padding(widget.Insets{
				Left:   8,
				Top:    8,
				Bottom: 8,
			}),
		)),
	)
	compositeContainer.AddChild(textAreaContainer)
	actionsBar.textArea = newTextArea("", widget.WidgetOpts.LayoutData(widget.RowLayoutData{
		Stretch:   false,
		MaxWidth:  200,
		MaxHeight: 600,
	}))
	textAreaContainer.AddChild(actionsBar.textArea)

	parentContainer := widget.NewContainer(
		widget.ContainerOpts.Layout(widget.NewRowLayout(
			widget.RowLayoutOpts.Spacing(-20),
			widget.RowLayoutOpts.Padding(widget.Insets{
				Left: 8,
				Top:  0,
			}),
		)),
	)
	compositeContainer.AddChild(parentContainer)

	transparentImg := archetype.NewTransparentEbitenImage(200, 3)
	transparentImageWidget := widget.NewGraphic(
		widget.GraphicOpts.Image(transparentImg),
	)
	messagesContainer := widget.NewContainer(
		widget.ContainerOpts.Layout(widget.NewStackedLayout()),
		widget.ContainerOpts.WidgetOpts(widget.WidgetOpts.LayoutData(widget.RowLayoutData{
			Position: widget.RowLayoutPositionStart,
			Stretch:  true,
		})),
	)

	parentContainer.AddChild(messagesContainer)
	messagesContainer.AddChild(transparentImageWidget)

	innerContainer := widget.NewContainer(
		widget.ContainerOpts.Layout(widget.NewRowLayout(
			widget.RowLayoutOpts.Spacing(-8),
			widget.RowLayoutOpts.Padding(widget.Insets{
				Left:   0,
				Bottom: 0,
			}),
		)),
	)
	messagesContainer.AddChild(innerContainer)

	inputContainer := widget.NewContainer(
		widget.ContainerOpts.Layout(widget.NewRowLayout(
			widget.RowLayoutOpts.Spacing(0),
			widget.RowLayoutOpts.Padding(widget.Insets{
				Left:   0,
				Bottom: 0,
			}),
		)),
	)
	innerContainer.AddChild(inputContainer)

	actionsBar.inputText = widget.NewTextInput(
		widget.TextInputOpts.WidgetOpts(
			//Set the layout information to center the textbox in the parent
			widget.WidgetOpts.LayoutData(widget.RowLayoutData{
				Position: widget.RowLayoutPositionCenter,
				Stretch:  false,
			}),
		),
		widget.TextInputOpts.Image(archetype.CreateRoundedTextInputImages(200, 50, 5, colornames.White, colornames.Fuchsia, colornames.Grey, 5)),
		widget.TextInputOpts.Placeholder(sendPlaceHolder),
		widget.TextInputOpts.Color(&widget.TextInputColor{
			Idle:          assets.BlueColor,
			Disabled:      colornames.Grey,
			Caret:         colornames.Gray,
			DisabledCaret: colornames.Grey,
		}),
		widget.TextInputOpts.Face(assets.MainFont),
		widget.TextInputOpts.CaretOpts(
			widget.CaretOpts.Color(colornames.Gray),
			widget.CaretOpts.Size(assets.MainFont, 16),
		),
		widget.TextInputOpts.ChangedHandler(func(args *widget.TextInputChangedEventArgs) {
			actionsBar.MessageValue = &args.TextInput.InputText
		}),
		widget.TextInputOpts.SubmitHandler(func(args *widget.TextInputChangedEventArgs) {
			if len(args.TextInput.InputText) > 0 {
				actionsBar.MessageValue = &args.TextInput.InputText
				actionsBar.MessageDone = true
				actionsBar.UpdateTextArea(fmt.Sprintf("\n[color=27BDF5]%s:[/color]", *actionsBar.Game.Session.UserName))
				actionsBar.UpdateTextArea(args.TextInput.InputText)
			}
		}),
		widget.TextInputOpts.Padding(widget.Insets{
			Top:    10,
			Bottom: 2,
			Left:   10,
			Right:  10,
		}),
		widget.TextInputOpts.RepeatInterval(150*time.Millisecond),
	)

	inputContainer.AddChild(actionsBar.inputText)

	sendButtonContainer, sendButton := archetype.CreateSVGImageButton(assets.SendKey, "Send", func() {
		if actionsBar.MessageValue != nil && len(*actionsBar.MessageValue) > 0 {
			actionsBar.MessageDone = true
		}
	})
	sendButtonContainer.GetWidget().LayoutData = widget.RowLayoutData{
		Position: widget.RowLayoutPositionEnd,
		Stretch:  false,
	}

	actionsBar.sendButton = sendButton
	innerContainer.AddChild(sendButtonContainer)

	buttonContainer := widget.NewContainer(
		widget.ContainerOpts.Layout(widget.NewRowLayout(
			widget.RowLayoutOpts.Spacing(-10),
			widget.RowLayoutOpts.Padding(widget.Insets{
				Left:   2,
				Bottom: 1,
			}),
		)),
		widget.ContainerOpts.WidgetOpts(widget.WidgetOpts.LayoutData(widget.AnchorLayoutData{
			HorizontalPosition: widget.AnchorLayoutPositionStart,
			VerticalPosition:   widget.AnchorLayoutPositionEnd,
		})),
	)

	closeButtonContainer, closeButton := archetype.CreateSVGImageButton(assets.CloseKey, "Close", func() {
		actionsBar.Close = true
	})
	shareButtonContainer, shareButton := archetype.CreateSVGImageButton(assets.ShareKey, "Copy ID", func() {
		actionsBar.Share = true
	})

	// Add the button containers to the row layout
	buttonContainer.AddChild(closeButtonContainer)
	buttonContainer.AddChild(shareButtonContainer)

	// Add the buttons to the AnchorLayout
	compositeContainer.AddChild(buttonContainer)

	actionsBar.container = container
	actionsBar.closeButton = closeButton
	actionsBar.shareButton = shareButton

	actionsBar.inputText.Focus(true)

	actionsBar.ui = &ebitenui.UI{
		Container: actionsBar.container,
	}

	return actionsBar
}

func newTextArea(text string, widgetOpts ...widget.WidgetOpt) *widget.TextArea {
	width := 64.0
	height := 64.0
	areaImage := ebiten.NewImage(int(width), int(height))
	borderWidth := 2.0
	aimg := archetype.DrawRoundedRect(areaImage, borderWidth, borderWidth, width-2*borderWidth, height-2*borderWidth, 1, colornames.White, assets.BlueColor, borderWidth)
	track := archetype.NewColoredEbitenImage(4, 40, archetype.ColorToRGBA(assets.BlueColor))
	handle := archetype.NewColoredEbitenImage(5, 5, archetype.ColorToRGBA(assets.BlueColor))
	lres := &listResources{
		image: &widget.ScrollContainerImage{
			Idle:     image.NewNineSlice(aimg, [3]int{12, 12, 12}, [3]int{16, 16, 32}),
			Disabled: image.NewNineSlice(aimg, [3]int{12, 12, 12}, [3]int{12, 12, 12}),
			Mask:     image.NewNineSlice(aimg, [3]int{12, 12, 12}, [3]int{12, 12, 12}),
		},

		track: &widget.SliderTrackImage{
			Idle:     image.NewNineSlice(track, [3]int{5, 0, 0}, [3]int{25, 12, 25}),
			Hover:    image.NewNineSlice(track, [3]int{5, 0, 0}, [3]int{25, 12, 25}),
			Disabled: image.NewNineSlice(track, [3]int{0, 5, 0}, [3]int{25, 12, 25}),
		},

		trackPadding: widget.Insets{
			Top:    5,
			Bottom: 24,
		},

		handle: &widget.ButtonImage{
			Idle:     image.NewNineSliceSimple(handle, 0, 5),
			Hover:    image.NewNineSliceSimple(handle, 0, 5),
			Pressed:  image.NewNineSliceSimple(handle, 0, 5),
			Disabled: image.NewNineSliceSimple(handle, 0, 5),
		},

		handleSize: 5,

		entryPadding: widget.Insets{
			Left:   8,
			Right:  8,
			Top:    2,
			Bottom: 8,
		},
	}
	return widget.NewTextArea(
		widget.TextAreaOpts.ContainerOpts(widget.ContainerOpts.WidgetOpts(widgetOpts...)),
		widget.TextAreaOpts.ScrollContainerOpts(widget.ScrollContainerOpts.Image(lres.image)),
		widget.TextAreaOpts.SliderOpts(
			widget.SliderOpts.Images(lres.track, lres.handle),
			widget.SliderOpts.MinHandleSize(lres.handleSize),
			widget.SliderOpts.TrackPadding(lres.trackPadding),
		),
		widget.TextAreaOpts.ShowVerticalScrollbar(),
		widget.TextAreaOpts.VerticalScrollMode(widget.PositionAtEnd),
		widget.TextAreaOpts.ProcessBBCode(true),
		widget.TextAreaOpts.FontFace(assets.MainSmallFont),
		widget.TextAreaOpts.FontColor(colornames.Black),
		widget.TextAreaOpts.TextPadding(lres.entryPadding),
		widget.TextAreaOpts.Text(text),
	)
}

func (s *ActionsBar) UpdateTextArea(text string) {
	nextText := s.textArea.GetText() + "\n" + text
	s.textArea.SetText(nextText)
}

func (s *ActionsBar) Reset() {
	s.inputText.InputText = ""
	s.MessageDone = false
	s.MessageValue = nil
}

func (s *ActionsBar) Draw(screen *ebiten.Image) {
	s.ui.Draw(screen)
}

func (s *ActionsBar) Container() *widget.Container {
	return s.container
}

func (s *ActionsBar) Update() {
	if s.World != nil {
		game := component.MustFindGame(*s.World)
		if game != nil {
			s.inputText.GetWidget().MinWidth = int(game.LeftOffset) - (20 + 32)
			s.ui.Container.GetWidget().LayoutData = widget.RowLayoutData{
				MaxWidth:  int(game.LeftOffset),
				MaxHeight: game.Settings.ScreenHeight,
			}
			textAreaWidget := s.textArea.GetWidget()
			textAreaWidget.MinWidth = int(game.LeftOffset) - (20)
			textAreaWidget.MinHeight = game.Settings.ScreenHeight - 114
			// s.textArea.MaxHeight = textAreaWidget.MinHeight
			// s.textArea.MaxWidth = textAreaWidget.MinWidth

			textAreaWidget.LayoutData = widget.WidgetOpts.LayoutData(widget.RowLayoutData{
				MaxWidth:  int(game.LeftOffset) - 20,
				MaxHeight: game.Settings.ScreenHeight - 114,
			})
			closeButtonRect := s.closeButton.GetWidget().Rect
			shareButtonRect := s.shareButton.GetWidget().Rect
			sendButtonRect := s.sendButton.GetWidget().Rect
			mx, my := ebiten.CursorPosition()
			if game.CursorOverButton {
				return
			}
			cursorOverButton := (closeButtonRect.Min.X <= mx && mx <= closeButtonRect.Max.X && closeButtonRect.Min.Y <= my && my <= closeButtonRect.Max.Y) ||
				(shareButtonRect.Min.X <= mx && mx <= shareButtonRect.Max.X && shareButtonRect.Min.Y <= my && my <= shareButtonRect.Max.Y) ||
				(sendButtonRect.Min.X <= mx && mx <= sendButtonRect.Max.X && sendButtonRect.Min.Y <= my && my <= sendButtonRect.Max.Y)

			if cursorOverButton != game.CursorOverButton {
				game.CursorOverButton = cursorOverButton
			}
		}
	}
	s.ui.Update()
}

Reply:
This is the code I'm running if you see, I have commented where I set the s.textArea.MaxHeight and s.textArea.MaxWidth and text area get too big without it, also game does not have enough data to set these values when UI get created, I have tried also to set the LayoutData of the widget and parent container here on Update, but I do get the same result.